### ISSUE TITLE
Stitching : gain compensation memory complexity is O(n²)

### ISSUE THREAD

**Issue by Tytan** [opencv/opencv#7470]
### System information (version)
- OpenCV => 3.1

### Related code : 
 ```.cpp
void GainCompensator::feed(const std::vector<Point> &corners, const std::vector<UMat> &images,
                           const std::vector<std::pair<UMat,uchar> > &masks)
{
    ...
    const int num_images = static_cast<int>(images.size());
    Mat_<int> N(num_images, num_images); N.setTo(0);
    Mat_<double> I(num_images, num_images); I.setTo(0);
```

### What is the problem ?
When using the GainCompensator, it is fine since the number of images is likely to be small.
Problems come with BlocksGainCompensator, which splits images into small block, and feeds these blocks to GainCompensator, resulting in a big number of images.

Let's do some practical maths here : 
I want to compensate gain on 3 images. 
The current stitching pipeline decimates my images to 0.1Mpx (by default), and uses blocks of 32x32 pixels (by default).
This gives (0.1e6 / (32*32) = 100) blocks per image. This is a total of 300 blocks.
I have to allocate a 300x300 matrix of double. This is 90 000 elements.

Now let's say I want to work on a full size image of 15Mpx.
This gives (15e6 / (32*32) = 15000) blocks per image. This is a total of 45000 blocks.
I have to allocate a 45000x45000 matrix of double. This is over 2e9 elements.

**=> Using the block gain compensation on big images, or using small block result in a memory allocation failure.**

### Why is it a problem ?
An answer to this problem is probably : "This is not a problem because we always work on decimated images". But I think this is taking the problem upside down. I think we decimate the image BECAUSE of that. 

The block-based approach is used to reduce complexity issues when working on big images, without having to decimate it. But in the OpenCV implementation, we use both image decimation and blocks-based gain compensation. 

For more information, read this reference : "Uyttendaele, A. Eden, and R. Szeliski. Eliminating ghosting and exposure artifacts in image mosaics. In Proc. CVPR’01, volume 2, pages 509–516, 2001".

### TL;DR
Using blocks gain compensation on big images, or using small blocks result in memory allocation failure.
Reducing the memory complexity of the algorithm will give more flexibility when choosing the decimated resolution and gain compensation block size.
